### PR TITLE
Use File.Open overload that overwrites file entirely in DiscoverCommand

### LINKED ISSUE #2104: TagHelpers.output.cache is not updated correctly
Repro:
- Create a simple app with a tag helper and RazorCompileOnBuild enabled
- Build and make sure `TagHelpers.output.cache` contains the serialized tag helper
- Delete the tag helper from the app
- Build again

Expected:
`TagHelper.output.cache` is empty

Actual:
`TagHelper.output.cache` still contains the deleted tag helper but the timestamp is updated

I am adding a test that covers this as we didn't have any.

cc @pranavkm @rynowak

### PR DESCRIPTION
**Pull request by pranavkm**
Re-enable TagHelper discovery incrementalism test

Fixes #2104

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.Tools/DiscoverCommand.cs [hunk 1]
```diff
@@ -150,15 +150,15 @@ private int ExecuteCore(RazorConfiguration configuration, string projectDirector
...
+                stream.Position = 0;
 
                 var newHash = Hash(stream);
                 var existingHash = Hash(outputFilePath);
 
                 if (!HashesEqual(newHash, existingHash))
                 {
                     stream.Position = 0;
-                    using (var output = File.OpenWrite(outputFilePath))
+                    using (var output = File.Open(outputFilePath, FileMode.Create))
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Razor.Tools/DiscoverCommand.cs]: OpenWrite does not reset the file length.

File: test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIncrementalismTest.cs [hunk 2]
```diff
@@ -56,5 +56,34 @@ public class BuildIncrementalismTest : MSBuildIntegrationTestBase
...
+            var fileThumbPrint = GetThumbPrint(generatedFile);
+
+            // Act - 2
+            // Update the source content and build. We should expect the outputs to be regenerated.
+            ReplaceContent(string.Empty, file);
+            result = await DotnetMSBuild("Build", "/p:RazorCompileOnBuild=true");
+
+            // Assert - 2
+            Assert.BuildPassed(result);
+            Assert.FileContentEquals(result, tagHelperOutputCache, "[]");
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIncrementalismTest.cs]: Verifying exact outputs now, the remainder of the test is the same

File: test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/RazorGenerateIntegrationTest.cs [hunk 3]
```diff
@@ -109,9 +109,6 @@ public class RazorGenerateIntegrationTest : MSBuildIntegrationTestBase
 
             // Act - 2
             // Update the source content and build. We should expect the outputs to be regenerated.
-            // Timestamps on xplat are precise only to a second. Add a delay so we can ensure that MSBuild recognizes the
-            // file change. See https://github.com/dotnet/corefx/issues/26024
-            await Task.Delay(TimeSpan.FromSeconds(1));
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/RazorGenerateIntegrationTest.cs]: Maybe this is sorted by now

File: test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIncrementalismTest.cs [hunk 4]
```diff
@@ -56,5 +56,34 @@ public class BuildIncrementalismTest : MSBuildIntegrationTestBase
                 }
             }
         }
+
+        [Fact]
+        [InitializeTestProject("SimpleMvc")]
+        public async Task Build_RegeneratesTagHelperInputs_IfFileChanges()
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIncrementalismTest.cs]: Still prefer naming this `RazorGenerate_Rebuilds_IfTagHelpersChanged` to make it consistent with others.

File: test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/RazorGenerateIntegrationTest.cs (hunk 3 above)
Comment (ajaybhargavb) [reply test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/RazorGenerateIntegrationTest.cs]: Looks like we don't have the fix yet. This test is failing on Travis.

File: src/Microsoft.AspNetCore.Razor.Tools/DiscoverCommand.cs (hunk 1 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Razor.Tools/DiscoverCommand.cs]: Oh. So does that mean that we could corrupt the contents?

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Razor.Tools/DiscoverCommand.cs]: Yup. It works fine as the content was the same length or longer.

File: test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/RazorGenerateIntegrationTest.cs (hunk 3 above)
Comment (pranavkm) [reply test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/RazorGenerateIntegrationTest.cs]: Oh well. The issue is marked as fixed, I was kinda hopeful.

File: test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIncrementalismTest.cs (hunk 4 above)
Comment (pranavkm) [reply test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIncrementalismTest.cs]: Sure

**Comment by pranavkm** [comment]
> flaky tests?

Sadly, AppVeyor issues that we haven't dug in to.

**Pull request merged**